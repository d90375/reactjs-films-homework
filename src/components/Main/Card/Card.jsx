import React from "react";
import PropTypes from "prop-types";
import "./Card.scss";

const Card = ({ posterImg, title, score, genresData, genres }) => {
  const imageNotFoundUrl = `url(https://www.themoviedb.org/assets/2/v4/glyphicons/basic/glyphicons-basic-38-picture-grey-c2ebdbb057f2a7614185931650f8cee23fa137b93812ccb132b9df511df1cfac.svg)`;

  const cardBackGroundStyle = {
    backgroundImage: posterImg ? `url(https://image.tmdb.org/t/p/w500${posterImg})` : imageNotFoundUrl
  };
  return (
    <div className="card">
      <div className="card__img" style={cardBackGroundStyle} />
      <div className="card__description">
        <div className="title">
          <span className="title__text">{title}</span>
          <div className="title__rating">
            <span className="title__number">{score}</span>
          </div>
        </div>
        <span className="card__description__genre">
          {genresData &&
            genres
              .reduce((acc, genre) => {
                return `${acc}${genresData[genre]}, `;
              }, "")
              .slice(0, -2)}
        </span>
      </div>
    </div>
  );
};

export default Card;

Card.propTypes = {
  posterImg: PropTypes.string,
  title: PropTypes.string.isRequired,
  genres: PropTypes.array.isRequired,
  genresData: PropTypes.object.isRequired,
  score: PropTypes.number.isRequired
};

Card.defaultProps = {
  posterImg: null
};
