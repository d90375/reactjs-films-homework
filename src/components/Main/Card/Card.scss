.card {
  width: 500px;
  height: 600px;
  background-color: #fff;
  margin-right: 40px;
  margin-bottom: 40px;
  border-radius: 5px;

  &:nth-child(4n) {
    margin-right: 0;
  }

  &:nth-child(n+9) {
    margin-bottom: 0;
  }
}

.card__img {
  height: 440px;
  background: url("../../../assets/images/mock/creed.jpg.png") right;
  background-size: cover;
  border-top-left-radius: 5px;
  border-top-right-radius: 5px;
}

.title {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: space-between;
  margin-bottom: 10px;
}

.title__text {
  color: #242b2d;
  font-family: "Helvetica Neue - bold", serif;
  font-size: 28px;
  font-weight: 700;
  line-height: 44px;
  text-transform: uppercase;
  letter-spacing: 1.12px;
}

.title__rating {
  text-align: center;
  width: 60px;
  height: 36px;
  border: 2px solid #49c1f0;
  border-radius: 5px;
}

.title__number {
  color: #47c3f0;
  font-family: "Helvetica Neue - bold", serif;
  font-size: 24px;
  font-weight: 500;
  line-height: 35px;
}

.card__description {
  margin-top: 40px;
  margin-left: 41px;
  margin-right: 41px;
}

.card__description__genre {
  color: #47c3f0;
  font-family: "Helvetica Neue - bold", serif;
  font-size: 22px;
  font-weight: 500;
  line-height: 33px;
  letter-spacing: 0.88px;
}
