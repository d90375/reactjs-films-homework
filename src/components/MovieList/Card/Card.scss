.card {
  width: 290px;
  height: 380px;
  position: relative;
  overflow: hidden;
  background-color: #fff;
  border-radius: 5px;
  margin-bottom: 1.5rem;

  &:nth-child(4n) {
    margin-right: 0;
  }

  &:nth-child(n+14) {
    margin-bottom: 5rem;
  }
}

.card__img {
  position: relative;
  border-radius: 5px 5px 0 0;
  height: calc(100% - 80px);
  transition: all 0.4s ease-in;
  background-size: cover;
  background-position: left top;
}

.card__title {
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  justify-content: space-between;
}

.card__title__text {
  width: 12rem;
  margin: 0;
  color: #242b2d;
  font-family: "Helvetica Neue - bold", serif;
  font-size: 18px;
  font-weight: 700;
  text-transform: uppercase;
  letter-spacing: 1.12px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.card__title__rating {
  text-align: center;
  width: 30px;
  height: 19px;
  border: 1px solid #49c1f0;
  border-radius: 5px;
}

.card__title__number {
  color: #47c3f0;
  font-family: "Helvetica Neue - Condensed Black", serif;
  font-size: 18px;
  font-weight: 500;
}

.card__description {
  padding: 1rem 1rem 0 1rem;
  cursor: pointer;
  height: 80px;
  outline: none;
}

.card__description__genre {
  color: #47c3f0;
  font-family: "Helvetica Neue - bold", serif;
  font-size: 0.7rem;
  font-weight: 500;
  letter-spacing: 0.88px;
}
