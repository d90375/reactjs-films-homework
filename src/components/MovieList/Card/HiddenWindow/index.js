import HiddenWindow from "./HiddenWindow";

export default HiddenWindow;
