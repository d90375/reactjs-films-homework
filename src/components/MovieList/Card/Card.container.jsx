import React, { useState, useCallback } from "react";
import PropTypes from "prop-types";
import { useDispatch } from "react-redux";

import Card from "./Card";
import { fetchMovieById } from "../../../modules/headerData";
import { fetchTrailerById } from "../../../modules/trailerData";

const CardContainer = React.memo(({ cardItem, genresItems, cardIndex }) => {
  const dispatch = useDispatch();

  const [isInfoShow, setIsInfoShow] = useState(false);

  const genres = cardItem?.genre_ids
    .reduce((acc, genre) => {
      return `${acc}${genresItems[genre]}, `;
    }, "")
    .slice(0, -2);

  const handleShowTrailer = useCallback(() => dispatch(fetchTrailerById(cardItem?.id)), [cardItem?.id, dispatch]);

  const handleChangeHeaderMovie = useCallback(() => {
    dispatch(fetchMovieById(cardItem?.id));
  }, [dispatch, cardItem?.id]);

  const handleShowInfo = useCallback(() => {
    setIsInfoShow(!isInfoShow);
  }, [isInfoShow]);

  return (
    <>
      <Card
        cardItem={cardItem}
        cardIndex={cardIndex}
        genres={genres}
        isInfoShow={isInfoShow}
        handleShowInfo={handleShowInfo}
        onChangeHeaderMovie={handleChangeHeaderMovie}
        handleShowTrailer={handleShowTrailer}
      />
    </>
  );
});

export default CardContainer;

Card.displayName = "Card";

CardContainer.propTypes = {
  cardItem: PropTypes.shape({
    popularity: PropTypes.number,
    vote_count: PropTypes.number,
    video: PropTypes.bool,
    poster_path: PropTypes.string,
    id: PropTypes.number,
    adult: PropTypes.bool,
    backdrop_path: PropTypes.string,
    original_language: PropTypes.string,
    original_title: PropTypes.string,
    genre_ids: PropTypes.arrayOf(PropTypes.number),
    title: PropTypes.string,
    vote_average: PropTypes.number,
    overview: PropTypes.string,
    release_date: PropTypes.string
  }),
  genresItems: PropTypes.shape({
    number: PropTypes.string
  }),
  cardIndex: PropTypes.number
};

CardContainer.defaultProps = {
  cardItem: {
    popularity: 0,
    vote_count: 0,
    video: false,
    poster_path: "",
    id: 0,
    adult: false,
    backdrop_path: "",
    original_language: "",
    original_title: "",
    genre_ids: [0, 0],
    title: "",
    vote_average: 0,
    overview: "",
    release_date: ""
  },
  genresItems: {
    0: ""
  },
  cardIndex: 0
};
