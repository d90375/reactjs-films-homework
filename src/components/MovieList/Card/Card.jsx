import React from "react";
import PropTypes from "prop-types";
import "./Card.scss";
import { IMAGE_NOT_FOUND_URL } from "../../../constants";
import HiddenWindow from "./HiddenWindow";
import InfoWindow from "./InfoWindow";

const Card = ({ cardItem, cardIndex, genres, isInfoShow, onChangeHeaderMovie, handleShowTrailer, handleShowInfo }) => {
  const { poster_path: posterImg, id, title, vote_average: score, overview } = cardItem;

  const cardImgStyle = {
    backgroundImage: posterImg ? `url(https://image.tmdb.org/t/p/w500${posterImg})` : `url(${IMAGE_NOT_FOUND_URL})`
  };

  return (
    <div className="card__wrap">
      {isInfoShow ? (
        <InfoWindow cardItem={{ id, title, score, genres, posterImg, overview }} onShowInfo={handleShowInfo} />
      ) : (
        <div className="card">
          <div style={cardImgStyle} className="card__img" />
          <button tabIndex={cardIndex} type="button" onClick={onChangeHeaderMovie} className="card__description">
            <div className="card__title">
              <h3 className="card__title__text">{title}</h3>
              <div className="card__title__rating">
                <span className="card__title__number">{score}</span>
              </div>
            </div>
            <span className="card__description__genre">{genres}</span>
          </button>
        </div>
      )}
      <HiddenWindow onShowInfo={handleShowInfo} onShowTrailer={handleShowTrailer} />
    </div>
  );
};

export default Card;

Card.propTypes = {
  cardItem: PropTypes.shape({
    id: PropTypes.number,
    posterImg: PropTypes.string,
    title: PropTypes.string,
    score: PropTypes.number,
    overview: PropTypes.string
  }),
  genres: PropTypes.string,
  cardIndex: PropTypes.number,
  isInfoShow: PropTypes.bool,
  onChangeHeaderMovie: PropTypes.func.isRequired,
  handleShowTrailer: PropTypes.func.isRequired,
  handleShowInfo: PropTypes.func.isRequired
};

Card.defaultProps = {
  cardItem: {
    id: 0,
    posterImg: "",
    title: "",
    score: 0,
    overview: ""
  },
  genres: "",
  cardIndex: 0,
  isInfoShow: false
};
