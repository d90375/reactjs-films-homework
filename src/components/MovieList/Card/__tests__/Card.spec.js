import React from "react";
import { shallow } from "enzyme";
import Card from "../Card";

describe("Card", () => {
  let component;

  const mockProps = {
    cardIndex: 9,
    genres: "genres",
    isInfoShow: true,
    isDisplayCardDirection: "line"
  };

  const mockCallBacks = {
    onChangeHeaderMovie: jest.fn(),
    handleShowTrailer: jest.fn(),
    handleShowInfo: jest.fn()
  };

  it("should render Card component without props", () => {
    component = shallow(<Card {...mockCallBacks} />);
    expect(component).toMatchSnapshot();
  });

  it("should render Card component with props", () => {
    component = shallow(
      <Card
        {...mockProps}
        cardItem={{ poster_path: "posterImg", title: "title", overview: "overview", vote_average: 9, id: 9 }}
        {...mockCallBacks}
      />
    );
    expect(component).toMatchSnapshot();
  });

  describe("should handler", () => {
    it("should call onChangeHeaderMovie method ", () => {
      component = shallow(<Card {...mockCallBacks} />);
      expect(mockCallBacks.onChangeHeaderMovie.mock.calls.length).toBe(0);
      component.find(".card__description").simulate("click");
      expect(mockCallBacks.onChangeHeaderMovie.mock.calls.length).toBe(1);
    });
  });
});
