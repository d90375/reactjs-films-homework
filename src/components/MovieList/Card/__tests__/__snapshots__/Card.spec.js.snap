// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should render Card component with props 1`] = `
<div
  className="card__wrap"
>
  <InfoWindow
    cardItem={
      Object {
        "genres": "genres",
        "id": 9,
        "overview": "overview",
        "posterImg": "posterImg",
        "score": 9,
        "title": "title",
      }
    }
    onShowInfo={[MockFunction]}
  />
  <HiddenWindow
    onShowInfo={[MockFunction]}
    onShowTrailer={[MockFunction]}
  />
</div>
`;

exports[`Card should render Card component without props 1`] = `
<div
  className="card__wrap"
>
  <div
    className="card"
  >
    <div
      className="card__img"
      style={
        Object {
          "backgroundImage": "url(https://www.themoviedb.org/assets/2/v4/glyphicons/basic/glyphicons-basic-38-picture-grey-c2ebdbb057f2a7614185931650f8cee23fa137b93812ccb132b9df511df1cfac.svg)",
        }
      }
    />
    <button
      className="card__description"
      onClick={[MockFunction]}
      tabIndex={0}
      type="button"
    >
      <div
        className="card__title"
      >
        <h3
          className="card__title__text"
        />
        <div
          className="card__title__rating"
        >
          <span
            className="card__title__number"
          />
        </div>
      </div>
      <span
        className="card__description__genre"
      />
    </button>
  </div>
  <HiddenWindow
    onShowInfo={[MockFunction]}
    onShowTrailer={[MockFunction]}
  />
</div>
`;
